integration/build/testdata/Dockerfile.testBuildPreserveOwnership

```diff
@@ -0,0 +1,57 @@
+# Set up files and directories with known ownership
+FROM busybox AS source
+RUN touch /file && chown 100:200 /file \
+ && mkdir -p /dir/subdir \
+ && touch /dir/subdir/nestedfile \
+ && chown 100:200 /dir \
+ && chown 101:201 /dir/subdir \
+ && chown 102:202 /dir/subdir/nestedfile
+
+FROM busybox AS test_base
+RUN mkdir -p /existingdir/existingsubdir \
+ && touch /existingdir/existingfile \
+ && chown 500:600 /existingdir \
+ && chown 501:601 /existingdir/existingsubdir \
+ && chown 501:601 /existingdir/existingfile
+
+
+# Copy files from the source stage
+FROM test_base AS copy_from
+COPY --from=source /file .
+# Copy to a non-existing target directory creates the target directory (as root), then copies the _contents_ of the source directory into it
+COPY --from=source /dir /dir
+# Copying to an existing target directory will copy the _contents_ of the source directory into it
+COPY --from=source /dir/. /existingdir
```

Comment:
I think there's a bug in the copy behavior (both in classic builder, and in buildkit); it's come up a few times, and matching with the `docker cp` documentation;

> `SRC_PATH` specifies a directory:
> 
> - `DEST_PATH` does not exist
>   - `DEST_PATH` is created as a directory and the contents of the source directory are copied into this directory
> - `DEST_PATH` exists and is a file
>   - Error condition: cannot copy a directory to a file
> - `DEST_PATH` exists and is a directory
>   - `SRC_PATH` does not end with `/.` (that is: slash followed by dot)
>     - the source directory is copied into this directory
>   - `SRC_PATH` does end with `/.` (that is: slash followed by dot)
>     - the content of the source directory is copied into this directory

So _without_ the trailing `/.`, this line should result in `dir` _itself_ being copied to `/existingdir/dir`; similar to `cp`:


Without trailing `/.`;

```bash
docker run --rm --workdir=/test alpine sh -c 'apk add --no-cache --quiet tree && mkdir -p src/dir/subdir dest/existingdir  && touch src/dir/subdir/nestedfile && cp -R src/dir dest/existingdir && cd /test/dest && tree'

.
└── existingdir
    └── dir
        └── subdir
            └── nestedfile

3 directories, 1 file
```

With trailing `/.`;

```bash
docker run --rm --workdir=/test alpine sh -c 'apk add --no-cache --quiet tree && mkdir -p src/dir/subdir dest/existingdir  && touch src/dir/subdir/nestedfile && cp -R src/dir/. dest/existingdir && cd /test/dest && tree'

.
└── existingdir
    └── subdir
        └── nestedfile

2 directories, 1 file
```

But doing COPY in a Dockerfile produces the same;


In a Dockerfile; without trailing `/.`;

```Dockerfile
DOCKER_BUILDKIT=0 docker build -<<EOF
FROM alpine AS source
RUN mkdir -p /dir/subdir && touch /dir/subdir/nestedfile

FROM alpine AS no_slash_dot
WORKDIR /dest
RUN apk add --no-cache --quiet tree
RUN mkdir existingdir
COPY --from=source /dir existingdir
RUN tree; exit 1
EOF
```

Produces;

```
.
└── existingdir
    └── subdir
        └── nestedfile

2 directories, 1 file
```


And with;


```bash
DOCKER_BUILDKIT=0 docker build -<<EOF
FROM alpine AS source
RUN mkdir -p /dir/subdir && touch /dir/subdir/nestedfile

FROM alpine AS slash_dot
WORKDIR /dest
RUN apk add --no-cache --quiet tree
RUN mkdir existingdir
COPY --from=source /dir/. existingdir
RUN tree; exit 1
EOF
```

Produces the same;

```
.
└── existingdir
    └── subdir
        └── nestedfile

2 directories, 1 file
```

Reply:
Also opened https://github.com/moby/buildkit/issues/791 for the above